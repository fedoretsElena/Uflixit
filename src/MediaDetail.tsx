import React, { Component } from 'react';
import axios from 'axios';

import ApiConfig from './core/ApiConfig';

import Loader from './Loader';
import ErrorMsg from './ErrorMsg';

import Movie from './models/Movie';
import TVShow from './models/TVShow';
import MediaFactory, { MediaType } from './models/MediaFactory';

interface IState {
    media: TVShow | Movie | any,
    loaded: boolean,
    loading: boolean,
    error: string | null
}

class MediaDetail extends Component<any> {
    state: IState = {
        media: {},
        error: null,
        loaded: false,
        loading: false
    };

    render() {
        const {creators, stars, keywords, yearStarted, yearEnded, summary, rating, title, image, genres} = this.state.media;

        return (
            <>
                {this.state.loading && !this.state.error && <Loader/>}
                {this.state.loaded &&
                <div className="row">
                    <div className="col col-4">
                        <img className="img-thumbnail mb-3"
                             src={image}
                             alt="game-poster"/>

                        {genres.map((genre: string, i: number) =>
                            <span key={i}
                                  className="badge badge-secondary mr-2 py-1 px-2 mt-1">{genre}</span>)}
                    </div>

                    <section className="col col-8">
                        <div className="d-flex justify-content-between align-items-center">
                            <h1 className="mb-0">{title}</h1>
                            <span className="badge badge-warning">{rating}</span>
                        </div>
                        <div className="mt-1 text-primary">{yearStarted} - {yearEnded}</div>

                        <div className="mt-5 w-75">
                            {summary}
                        </div>

                        <div className="mt-2 mb-2">
                            {keywords.map((word: string, i: number) =>
                                <span key={i}
                                      className="badge badge-info mr-2 py-1 px-2 mt-1">{word}</span>)}
                        </div>

                        <div className="row mt-4">
                            {creators && creators.length > 0 && <div className="col col-6">
                                <h5>Creators:</h5>
                                <ul className="list-group list-group-flush">
                                    {creators.map((creator: string, i: number) =>
                                        <li key={i}
                                            className="list-group-item">{creator}</li>)}
                                </ul>
                            </div>}

                            <div className="col col-6">
                                <h5>Stars</h5>
                                <ul className="list-group list-group-flush">
                                    {stars.map((star: string, i: number) =>
                                        <li key={i}
                                            className="list-group-item">{star}</li>)}
                                </ul>
                            </div>
                        </div>
                    </section>
                </div>}

                {this.state.error && <ErrorMsg msg={this.state.error}/>}
            </>
        );
    }

    componentDidMount(): void {
        this.getShowDetail();
    }

    private getShowDetail(): void {
        const {id} = this.props.match.params;
        const type = this.props.location.pathname.indexOf('movie') !== -1 ? MediaType.Movie : MediaType.TVShow;

        this.setState({
            loading: true
        });

        axios.get(ApiConfig.getMediaDetailsPath
            .replace('{id}', id)
            .replace('{type}', type))
            .then(data => {
                const factory = new MediaFactory(type, data);

                this.setState({
                    error: null,
                    loaded: true,
                    loading: false,
                    media: factory.getMedia()
                });
            })
            .catch(() => {
                this.setState({
                    error: 'Can not get details by id.'
                });
            });
    }
}

export default MediaDetail;