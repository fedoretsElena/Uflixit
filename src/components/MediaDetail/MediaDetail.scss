.col {
    &--perspective {
        perspective: 1600px;
    }
}

.cover-bg {
    background-size: cover;
    background-position: center;
    background-repeat: no-repeat;
}

.media__bg {
    top: 0;
    left: 0;
    height: 40vh;
    z-index: -1;
    opacity: 0.22;
    background-size: 100%;
    background-position: top;
    background-repeat: no-repeat;

    &:after {
        top: 31vh;
        content: "";
        width: 100%;
        height: 6em;
        background: linear-gradient(to bottom, transparent, #fff);
        position: absolute;
    }
}
