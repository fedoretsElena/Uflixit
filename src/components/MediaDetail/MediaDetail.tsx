import React, { Component } from "react";
import { CSSTransition } from "react-transition-group";

import ErrorMsg from "../shared/ErrorMsg";
import ListGroup from "../shared/ListGroup";
import Loader from "../shared/Loader";
import Rating from "../shared/Rating";

import Movie from "../../models/Movie";
import TVShow from "../../models/TVShow";
import MediaService from "../../data/services/mediaService";

import "./MediaDetail.scss";

interface IState {
    media: TVShow | Movie | any;
    loaded: boolean;
    loading: boolean;
    error: string | null;
}

class MediaDetail extends Component<any, IState> {
    state: IState = {
        media: {},
        error: null,
        loaded: false,
        loading: false
    };

    mediaService = new MediaService();

    render() {
        const {
            year,
            creators,
            stars,
            keywords,
            yearStarted,
            yearEnded,
            summary,
            rating,
            title,
            image,
            genres,
            fanart
        } = this.state.media;
        const bgStyle = {
            backgroundImage: `url(${fanart})`
        };

        return (
            <>
                {this.state.loading && !this.state.error && <Loader />}
                {this.state.loaded && (
                    <div className="row mt-5 mb-5">
                        <div
                            className="media__bg cover-bg position-absolute w-100 h-100"
                            style={bgStyle}
                        />
                        <div className="col col-4 col--perspective">
                            <CSSTransition
                                in={!!image}
                                timeout={300}
                                appear={true}
                                classNames="fadeRotate"
                            >
                                <img
                                    className="d-block img-thumbnail mb-4"
                                    src={image}
                                    alt="media-poster"
                                />
                            </CSSTransition>

                            {genres.map((genre: string, i: number) => (
                                <span
                                    key={i}
                                    className="badge badge-secondary mr-2 py-1 px-2 mt-1"
                                >
                                    {genre}
                                </span>
                            ))}
                        </div>

                        <section className="col col-8">
                            <div className="d-flex justify-content-between align-items-center">
                                <h1 className="mb-0 w-75">{title}</h1>
                                <Rating value={rating} />
                            </div>
                            <div className="mt-1 text-primary">
                                {" "}
                                {yearStarted ? (
                                    <span>
                                        {yearStarted} - {yearEnded}
                                    </span>
                                ) : (
                                    year
                                )}{" "}
                            </div>

                            <div className="mt-5 w-75">{summary}</div>

                            <div className="mt-2 mb-2">
                                {keywords.map((word: string, i: number) => (
                                    <span
                                        key={i}
                                        className="badge badge-info mr-2 py-1 px-2 mt-1"
                                    >
                                        {word}
                                    </span>
                                ))}
                            </div>

                            <div className="row mt-4">
                                <div className="col col-6">
                                    {creators && creators.length > 0 && (
                                        <ListGroup
                                            title="Creators"
                                            list={creators}
                                        />
                                    )}
                                </div>

                                <div className="col col-6">
                                    {stars && stars.length > 0 && (
                                        <ListGroup title="Stars" list={stars} />
                                    )}
                                </div>
                            </div>
                        </section>

                        <section className="col-12 mt-4"> test</section>
                    </div>
                )}

                {this.state.error && <ErrorMsg msg={this.state.error} />}
            </>
        );
    }

    componentDidMount(): void {
        this.getDetail();
    }

    private getDetail(): void {
        this.setState({
            loading: true
        });

        this.mediaService
            .getDetail(this.props.match.params.id, this.props.location)
            .then((data: Movie | TVShow) => {
                this.setState({
                    error: null,
                    loaded: true,
                    loading: false,
                    media: data
                });
            })
            .catch(() => {
                this.setState({
                    error: "Can not get details by id."
                });
            });
    }
}

export default MediaDetail;
