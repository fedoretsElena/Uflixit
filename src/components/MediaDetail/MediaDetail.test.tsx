import MediaDetail from "./MediaDetail";

jest.mock("../../data/services/mediaService");

describe("MediaDetail component", () => {
    let mockProps: { match: any; location: any };

    beforeEach(() => {
        mockProps = {
            match: {
                params: {
                    id: "1"
                }
            },
            location: {
                pathname: "https://uflixit/tv-shows/"
            }
        };
    });

    it("should render correctly and take data", done => {
        const component = shallow(
            <MediaDetail
                match={mockProps.match}
                location={mockProps.location}
            />
        );

        expect(component).toMatchSnapshot();

        setTimeout(() => {
            component.update();
            expect(component.instance().state.media).toBeDefined();
            done();
        });
    });

    it("should insert error component after fail request", () => {
        const component = mount(
            <MediaDetail
                match={mockProps.match}
                location={mockProps.location}
            />
        );

        component
            .instance()
            .setState({ error: "Error appear", loaded: false, loading: false }); // it's really bad idea

        const errorComponent = component.find("ErrorMsg");
        expect(errorComponent).toBeDefined();
    });
});
