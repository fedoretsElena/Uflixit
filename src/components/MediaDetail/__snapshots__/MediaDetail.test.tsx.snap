// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaDetail component should render correctly and take data 1`] = `
<Fragment>
  <Loader />
</Fragment>
`;
